### PR TITLE
Pashah/more rest commands

### PR DESCRIPTION
**Pull request by parthsha**
Changes proposed in this pull request:
 - Updated Out-RsRestCatalogItem and  Write-RsRestCatalogItem to use the official REST API 
 - Created new commands: Get-RsRestItemDataSources, New-RsRestFolder, Out-RsRestFolderContent, Remove-RsRestFolder, Remove-RsRestCatalogItem, Set-RsRestItemDataSources, and Write-RsRestFolderContent

How to test this code:
 - Added unit tests for all the new REST commands and the earlier REST commands

Has been tested on (remove any that don't apply):
 - Powershell 3 and above
 - Windows 7 and above
 - SQL Server 2012 and above

### CODE REVIEW

File: ReportingServicesTools/Functions/CatalogItems/Rest/Get-RsRestItemDataSources.ps1 [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+            Description
+            -----------
+            Fetches data sources associated to "MyReport" catalog item found in "/" folder.
+    #>
+
+    [CmdletBinding(SupportsShouldProcess = $true, ConfirmImpact = 'Medium')]
+    param(
+        [Parameter(Mandatory = $True)]
+        [Alias('ItemPath','Path')]
```
Comment (jtarquino) [inline ReportingServicesTools/Functions/CatalogItems/Rest/Get-RsRestItemDataSources.ps1]: Alias would be optional for new functions, at least you want to keep it in parity with the non rest version

File: ReportingServicesTools/Functions/CatalogItems/Rest/Get-RsRestItemDataSources.ps1 [hunk 2]
```diff
@@ -0,0 +1,103 @@
+# Copyright (c) 2016 Microsoft Corporation. All Rights Reserved.
+# Licensed under the MIT License (MIT)
+
+function Get-RsRestItemDataSources
```
Comment (jtarquino) [inline ReportingServicesTools/Functions/CatalogItems/Rest/Get-RsRestItemDataSources.ps1]: Our functions are typically singular (DataSource

File: ReportingServicesTools/Functions/CatalogItems/Rest/Out-RsRestCatalogItem.ps1 [hunk 3]
```diff
@@ -5,10 +5,10 @@ function Out-RsRestCatalogItem
 {
     <#
         .SYNOPSIS
-            This command downloads catalog items from a report server to disk. It is for SQL Server Reporting Service 2016 and later.
+            This command downloads catalog items from a report server to disk (using REST endpoint). It is for SQL Server Reporting Service 2016 and later.
```
Comment (jtarquino) [inline ReportingServicesTools/Functions/CatalogItems/Rest/Out-RsRestCatalogItem.ps1]: As this uses v2 as default I would change it to say 2017 (and add the note for 2016 use v1)

File: ReportingServicesTools/Functions/CatalogItems/Rest/Out-RsRestFolderContent.ps1 [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        }
+
+        $catalogItems = (ConvertFrom-Json $response.Content).value
+        foreach ($catalogItem in $catalogItems)
+        {
+            if ($catalogItem.Type -eq "Report" -or
+                $catalogItem.Type -eq "DataSet" -or
+                $catalogItem.Type -eq "DataSource" -or
+                $catalogItem.Type -eq "MobileReport" -or
+                $catalogItem.Type -eq "PowerBIReport")
```
Comment (jtarquino) [inline ReportingServicesTools/Functions/CatalogItems/Rest/Out-RsRestFolderContent.ps1]: not exporting resources?

File: ReportingServicesTools/Functions/CatalogItems/Rest/Out-RsRestFolderContent.ps1 [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+                Out-RsRestCatalogItemId -RsItemInfo $catalogItem -Destination $Destination -ReportPortalUri $ReportPortalUri -RestApiVersion $RestApiVersion -Credential $Credential -WebSession $WebSession
+            }
+            elseif ($catalogItem.Type -eq "Folder" -and $Recurse)
+            {
+                $subFolderPath = "$Destination\$($catalogItem.Name)"
+                Write-Verbose "Creating folder $($catalogItem.Name)..."
+                New-Item -Path $subFolderPath -ItemType Directory
+
+                Out-RsRestFolderContent -RsFolder $catalogItem.Path -Destination $subFolderPath -ReportPortalUri $ReportPortalUri -RestApiVersion $RestApiVersion -Credential $Credential -WebSession $WebSession
+            }
```
Comment (jtarquino) [inline ReportingServicesTools/Functions/CatalogItems/Rest/Out-RsRestFolderContent.ps1]: I would add a write-verbose with the items that aren't being exported

File: ReportingServicesTools/Functions/CatalogItems/Rest/Set-RsRestItemDataSources.ps1 [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            Specify the version of REST Endpoint to use. Valid values are: "v2.0".
+
+        .PARAMETER Credential
+            Specify the credentials to use when connecting to the Report Server.
+
+        .PARAMETER WebSession
+            Specify the session to be used when making calls to REST Endpoint.
+
+        .EXAMPLE
+            Set-RsRestItemDataSources -RsItem $myReport
```
Comment (jtarquino) [inline ReportingServicesTools/Functions/CatalogItems/Rest/Set-RsRestItemDataSources.ps1]: probablyu should add an example of how to get $myReport and how to assign the properties

**Comment by parthsha** [comment]
Thanks @jtarquino. Sorry for making you do this during the weekend.

File: ReportingServicesTools/Functions/CatalogItems/Rest/Out-RsRestCatalogItem.ps1 (hunk 3 above)
Comment (parthsha) [reply ReportingServicesTools/Functions/CatalogItems/Rest/Out-RsRestCatalogItem.ps1]: Great point. Removed "It is for SQL Server Reporting Services 2016 and later" and added use "V1 for 2016 under the parameter description for RestApiVersion.

File: ReportingServicesTools/Functions/CatalogItems/Rest/Out-RsRestFolderContent.ps1 (hunk 4 above)
Comment (parthsha) [reply ReportingServicesTools/Functions/CatalogItems/Rest/Out-RsRestFolderContent.ps1]: Added support for Resources. We didn't have support for it in the existing Out-RsFolderContent, Out-RsCatalogItem, Write-RsFolderContent and Write-RsCatalogItem.

File: ReportingServicesTools/Functions/CatalogItems/Rest/Out-RsRestFolderContent.ps1 (hunk 5 above)
Comment (parthsha) [reply ReportingServicesTools/Functions/CatalogItems/Rest/Out-RsRestFolderContent.ps1]: Added support so don't need to do this anymore.

File: ReportingServicesTools/Functions/CatalogItems/Rest/Out-RsRestFolderContent.ps1 (hunk 4 above)
Comment (parthsha) [reply ReportingServicesTools/Functions/CatalogItems/Rest/Out-RsRestFolderContent.ps1]: Also added support for Excel Workbook.

File: ReportingServicesTools/Functions/CatalogItems/Rest/Set-RsRestItemDataSources.ps1 (hunk 6 above)
Comment (parthsha) [reply ReportingServicesTools/Functions/CatalogItems/Rest/Set-RsRestItemDataSources.ps1]: I agree. I was looking at this API and the amount of knowledge required in just setting the data source credentials makes me feel that more work should be put in to make this command usable, if not useful. I am going to pull this and Get-RsItemDataSources out of this PR and check them in later.

File: ReportingServicesTools/Functions/CatalogItems/Rest/Get-RsRestItemDataSources.ps1 (hunk 2 above)
Comment (parthsha) [reply ReportingServicesTools/Functions/CatalogItems/Rest/Get-RsRestItemDataSources.ps1]: Removed this command for now. It was required for Set-RsRestItemDataSources, which I am pulling out from this PR.

**Comment by jzabroski** [comment]
@jtarquino  This was a breaking change:

```
        default { return 'Resource' }
```
was:
```
        default {throw 'Currently only .rdl, .rds, .rsds, .rsd and .rsmobile files are supported'}
```

Now if we release SQL Server 2025 and somebody is still in ReportingServicesTools 0.0.4.7 in 7 years, you will have 7 years of bad luck. :)

**Pull request merged**